larsimrad/BxDecay0/BaseRadioGen.h

```diff
@@ -172,105 +265,152 @@ namespace evgen {
       m_T1 = -m_T0;
     }
 
-    m_geo_volume_mode = pset.get_if_present<std::string>("volume_rand", m_volume_rand);
-
     m_geo_manager = m_geo_service->ROOTGeoManager();
 
-    if (not m_geo_volume_mode) {
+    m_random_flat    = std::make_unique<CLHEP::RandFlat   >(m_engine);
+    m_random_poisson = std::make_unique<CLHEP::RandPoisson>(m_engine);
+
+    if (m_material != ".*" || m_volume_gen != ".*") {
+      std::cout << "Calculating the volume of " << m_material << " and the volume " << m_volume_gen << " in the geometry.\n";
+      
+      double xyz[3];
+
+      TObjArray* volumes   = m_geo_manager->GetListOfVolumes();
+      TList    * materials = m_geo_manager->GetListOfMaterials();
+      
+      for (int i=0; i<volumes->GetEntries(); ++i) {
+        TGeoVolume* volume = (TGeoVolume*)volumes->At(i);
+        std::string volume_name = volume->GetName();
+        bool good = std::regex_match(volume_name, m_regex_volume);
+        if (good) {
+          
+          if (m_volume_gen != ".*") {
+            std::cout << "  Volume " << volume_name << " is an accepted volume of " << volume->GetShape()->Capacity() << "cm^3.\n";
+            if (volume->GetShape()->TestShapeBits(TGeoShape::kGeoBox)) {
+              TGeoBBox* box = dynamic_cast<TGeoBBox*>(volume->GetShape());
+              if (box)
+                std::cout << "  It is a box of size " << 2.*box->GetDX() << " x " << 2.*box->GetDY() << " x " << 2.*box->GetDZ() << " cm^3\n";
+            }
+          }
+          
+          m_good_volumes.push_back(volume);
+        }
+      }
+      
+      for (int i=0; i<materials->GetEntries(); ++i) {
+        TGeoMaterial* material = (TGeoMaterial*)materials->At(i);
+        std::string material_name = material->GetName();
+        bool good = std::regex_match(material_name, m_regex_material);
+        if (good) m_good_materials.push_back(material);
+      }
+    }
+    std::cout << m_good_volumes  .size() << " volumes correspond to the regex \""   << m_volume_gen << "\".\n";
+    std::cout << m_good_materials.size() << " materials correspond to the regex \"" << m_material   << "\".\n";
+    
+    double dummy;
+    if (pset.get_if_present<double>("X0", dummy) or
+        pset.get_if_present<double>("X1", dummy) or
+        pset.get_if_present<double>("Y0", dummy) or
+        pset.get_if_present<double>("Y1", dummy) or
+        pset.get_if_present<double>("Z0", dummy) or
+        pset.get_if_present<double>("Z1", dummy)) {
+      /// If we specified X, Y, Z we will throw with what the user provided
+      /// This will throw an error if somebody specifies only one X0.
+      
+      m_geo_volume_mode = false;
+      m_volume_rand_present = false;
+      
       m_X0 = pset.get<double>("X0");
       m_Y0 = pset.get<double>("Y0");
       m_Z0 = pset.get<double>("Z0");
       m_X1 = pset.get<double>("X1");
       m_Y1 = pset.get<double>("Y1");
       m_Z1 = pset.get<double>("Z1");
```

Comment:
I would not use a `dummy` variable here.  There are a couple other options:

```c++
if (pset.has_key("X0") or pset.has_key("X1") or
    pset.has_key("Y0") or pset.has_key("Y1") or
    pset.has_key("Z0") or pset.has_key("Z1")) 
{
  /// This will throw an error if somebody specifies only one X0.
 
  m_geo_volume_mode = false;
  m_volume_rand_present = false;
 
  m_X0 = pset.get<double>("X0");
  m_Y0 = pset.get<double>("Y0");
  m_Z0 = pset.get<double>("Z0");
  m_X1 = pset.get<double>("X1");
  m_Y1 = pset.get<double>("Y1");
  m_Z1 = pset.get<double>("Z1");
   ...
}
```

or a perhaps better option would be:

```c++
size_t n_params = 0;
n_params += static_cast<int>(pset.get_if_present("X0", m_X0));
n_params += static_cast<int>(pset.get_if_present("Y0", m_Y0));
n_params += static_cast<int>(pset.get_if_present("Z0", m_Z0));
n_params += static_cast<int>(pset.get_if_present("X1", m_X1));
n_params += static_cast<int>(pset.get_if_present("Y1", m_Y1));
n_params += static_cast<int>(pset.get_if_present("Z1", m_Z1));

if (n_params > 0 and n_params < 6) {
  throw exception(...) << "All or none of the following parameters must be specified"
}

m_geo_volume_mode = false;
m_volume_rand_present = false;
```

Reply:
done!